lib/store/session/webstorage.js

```diff
@@ -0,0 +1,143 @@
+     */
+    getEndToEndAccount: function() {
+        return this.store.getItem(KEY_END_TO_END_ACCOUNT);
+    },
+
+    /**
+     * Stores the known devices for a user.
+     * @param {string} userId The user's ID.
+     * @param {object} devices A map from device ID to keys for the device.
+     */
+    storeEndToEndDevicesForUser: function(userId, devices) {
+        setJsonItem(this.store, keyEndToEndDevicesForUser(userId), devices);
+    },
+
+    /**
+     * Retrieves the known devices for a user.
+     * @param {string} userId The user's ID.
+     * @return {object} A map from device ID to keys for the device.
+     */
+    getEndToEndDevicesForUser: function(userId)  {
+        return getJsonItem(this.store, keyEndToEndDevicesForUser(userId));
+    },
+
+    /**
+     * Store a session between the logged-in user and another device
+     * @param {string} deviceKey The public key of the other device.
+     * @param {string} sessionId The ID for this end-to-end session.
+     * @param {string} session Base64 encoded end-to-end session.
+     */
+    storeEndToEndSession: function(deviceKey, sessionId, session) {
+        var sessions = this.getEndToEndSessions(deviceKey) || {};
+        sessions[sessionId] = session;
+        setJsonItem(
+            this.store, keyEndToEndSessions(deviceKey), sessions
+        );
+    },
+
+    /**
+     * Retrieve the end-to-end sessions between the logged-in user and another
+     * device.
+     * @param {string} deviceKey The public key of the other device.
+     * @return {object} A map from sessionId to Base64 end-to-end session.
+     */
+    getEndToEndSessions: function(deviceKey) {
+        return getJsonItem(this.store, keyEndToEndSessions(deviceKey));
+    },
+
+    /**
+     * Store the end-to-end state for a room.
+     * @param {string} roomId The room's ID.
+     * @param {object} roomInfo The end-to-end info for the room.
+     */
+    storeEndToEndRoom: function(roomId, roomInfo) {
+        setJsonItem(this.store, keyEndToEndRoom(roomId), roomInfo);
+    },
+
+    /**
+     * Get the end-to-end state for a room
+     * @param {string} roomId The room's ID.
+     * @return {object} The end-to-end info for the room.
+     */
+    getEndToEndRoom: function(roomId) {
+        return getJsonItem(this.store, keyEndToEndRoom(roomId));
+    }
+};
+
+var KEY_END_TO_END_ACCOUNT = E2E_PREFIX + "account";
+
+function keyEndToEndDevicesForUser(userId) {
+    return E2E_PREFIX + "devices/" + userId;
+}
+
+function keyEndToEndSessions(deviceKey) {
+    return E2E_PREFIX + "sessions/" + deviceKey;
+}
+
+function keyEndToEndRoom(roomId) {
+    return E2E_PREFIX + "rooms/" + roomId;
+}
+
+function getJsonItem(store, key) {
+    try {
+        return JSON.parse(store.getItem(key));
+    }
+    catch (e) {
+        debuglog("Failed to get key %s: %s", key, e);
+        debuglog(e.stack);
+    }
+    return null;
+}
+
+function setJsonItem(store, key, val) {
+    store.setItem(key, JSON.stringify(val));
```

Comment:
Probably need to handle running out of space here (nb. Safari in privte browsing mode will always throw an exception when asked to save anything to localstorage, so I expect this will just break).